lib/improver/nowcasting/pysteps_advection.py

```diff
@@ -0,0 +1,249 @@
+                interval (int):
+                    Lead time interval, in minutes
+                gridlength (float):
+                    Size of grid square, in metres
+
+            Returns:
+                np.ndarray:
+                    Array of displacements in grid squares per time step
+            """
+            cube_ms = cube.copy()
+            cube_ms.convert_units('m s-1')
+            displacement = cube_ms.data*interval*60. / gridlength
+            return np.ma.filled(displacement, np.nan)
+
+        gridlength = calculate_grid_spacing(self.analysis_cube, 'metres')
+        udisp = _calculate_displacement(ucube, self.interval, gridlength)
+        vdisp = _calculate_displacement(vcube, self.interval, gridlength)
+        displacement = np.array([udisp, vdisp])
+        return displacement
+
+    def _reformat_analysis_cube(self):
+        """
+        Add forecast reference time and forecast period coordinates (if they do
+        not already exist) and nowcast attributes to analysis cube
+        """
+        coords = [coord.name() for coord in self.analysis_cube.coords()]
+        if "forecast_reference_time" not in coords:
+            frt_coord = self.analysis_cube.coord('time').copy()
+            frt_coord.rename('forecast_reference_time')
+            self.analysis_cube.add_aux_coord(frt_coord)
+        if "forecast_period" not in coords:
+            self.analysis_cube.add_aux_coord(
+                AuxCoord(np.array([0], dtype=np.int32),
+                         'forecast_period', 'seconds'))
+        # set nowcast attributes
+        self.analysis_cube.attributes['source'] = 'MONOW'
+        self.analysis_cube.attributes['title'] = (
+            'MONOW Extrapolation Nowcast on UK 2 km Standard Grid')
+
+    def _set_up_output_cubes(self, all_forecasts):
+        """
+        Convert 3D numpy array into list of cubes with correct time metadata.
+
+        Args:
+            all_forecasts (np.ndarray):
+                Array of 2D forecast fields returned by extrapolation function
+
+        Returns:
+            forecast_cubes (iris.cube.CubeList):
+                List of extrapolated cubes with correct time coordinates
+        """
+        current_datetime = iris_time_to_datetime(
+            self.analysis_cube.coord('time'))[0]
+        forecast_cubes = [self.analysis_cube.copy()]
+        for i in range(len(all_forecasts)):
+            # copy forecast data into template cube
+            new_cube = self.analysis_cube.copy(
+                data=all_forecasts[i, :, :].astype(np.float32))
+            # update time and forecast period coordinates
+            current_datetime += timedelta(seconds=self.interval*60)
+            current_time = datetime_to_iris_time(current_datetime)
+            new_cube.coord('time').points = np.array(
+                [current_time], dtype=np.int64)
+            new_cube.coord('forecast_period').points = np.array(
+                [(i+1)*self.interval*60], dtype=np.int32)
```

Comment:
Use forecast_period_coord function here too.
I think this is the only purpose for adding a forecast_period coord to the analysis cube, in which case you don't need it on the analysis cube if you change this to use the forecast_period_coord function. I hope that made sense...

Reply:
The reason for adding a forecast period coordinate to the analysis cube is that the updated analysis is actually returned as the T+0 forecast (first item in the cube list), at which point we want it to have a forecast reference time and forecast period coordinate consistent with all the later forecasts in the list.

I'm afraid I don't see the problem with adding point values here.  It doesn't use the centralised functionality, but the code tells a "story" in the sense that the time and forecast period are being updated together, in sequence, for each forecast.  It makes sense to me!